apisix/discovery/nacos.lua

```diff
@@ -0,0 +1,237 @@
+        timeout = {
+            type = "object",
+            properties = {
+                connect = {type = "integer", minimum = 1, default = 2000},
+                send = {type = "integer", minimum = 1, default = 2000},
+                read = {type = "integer", minimum = 1, default = 5000},
+            }
+        },
+    },
+    required = {"host"}
+}
+
+
+local _M = {
+    version = 0.1,
+}
+
+
+local function service_info()
+    local host = local_conf.discovery and
+            local_conf.discovery.nacos and local_conf.discovery.nacos.host
+    if not host then
+        log.error("do not set nacos.host")
+        return
+    end
+
+    local basic_auth
+    -- TODO Add health check to get healthy nodes.
+    local url = host[math_random(#host)]
+    local auth_idx = str_find(url, "@")
+    if auth_idx then
+        local protocol_idx = str_find(url, "://")
+        local protocol = string_sub(url, 1, protocol_idx + 2)
+        local user_and_password = string_sub(url, protocol_idx + 3, auth_idx - 1)
+        local other = string_sub(url, auth_idx + 1)
+        url = protocol .. other
+        basic_auth = "Basic " .. ngx.encode_base64(user_and_password)
+    end
+    if local_conf.discovery.nacos.prefix then
+        url = url .. local_conf.discovery.nacos.prefix
+    end
+    if string_sub(url, #url) ~= "/" then
+        url = url .. "/"
+    end
+
+    return url, basic_auth
+end
+
+
+local function request(request_uri, basic_auth, method, path, query, body)
+    log.info("nacos uri:", request_uri, ".")
+    local url = request_uri .. path
+    local headers = core.table.new(0, 5)
+    headers['Connection'] = 'Keep-Alive'
+    headers['Accept'] = 'application/json'
+
+    if basic_auth then
+        headers['Authorization'] = basic_auth
+    end
+
+    if body and 'table' == type(body) then
+        local err
+        body, err = core.json.encode(body)
+        if not body then
+            return nil, 'invalid body : ' .. err
+        end
+        -- log.warn(method, url, body)
+        headers['Content-Type'] = 'application/json'
+    end
+
+    local httpc = http.new()
+    local timeout = local_conf.discovery.nacos.timeout
+    local connect_timeout = timeout and timeout.connect or 2000
+    local send_timeout = timeout and timeout.send or 2000
+    local read_timeout = timeout and timeout.read or 5000
+    log.info("connect_timeout:", connect_timeout, ", send_timeout:", send_timeout,
+        ", read_timeout:", read_timeout, ".")
+    httpc:set_timeouts(connect_timeout, send_timeout, read_timeout)
+    return httpc:request_uri(url, {
+        version = 1.1,
+        method = method,
+        headers = headers,
+        query = query,
+        body = body,
+        ssl_verify = false,
+    })
+end
+
+
+local function get_url(request_url,basic_auth,path)
+    local res, err = request(request_url, basic_auth, "GET", path)
+    if not res then
+        log.error("failed to fetch registry", err)
+        return
```

Comment:
Please pass the error out so the caller can check it.

Reply:
OK.
By the way,when i add nacos test script in centos7-ci.yml,can't submit,how can i  fix it?
error tips:
[remote rejected] (refusing to allow an OAuth App to create or update workflow `.github/workflows/centos7-ci.yml` without `workflow` scope)

Reply:
content:
```
 rm -rf nacos-docker
        git clone --depth 1 https://github.com/benx203/nacos-docker.git
        cd nacos-docker
        docker-compose -f example/standalone-test.yaml up
        ps aux|grep nacos-gateway-provider-example-2.2.6-SNAPSHOT |grep -v grep|awk '{print $2}'|xargs kill -9
        nohup java -jar example/nacos-gateway-provider-example-2.2.6-SNAPSHOT.jar --spring.application.name=APISIX-NACOS  > /dev/null 2>&1 &
```

Reply:
> refusing to allow an OAuth App

Try to update it locally via a regular editor and submit it with Git?

Reply:
Thanks,Submit it with TortoiseGit success.